### ISSUE TITLE
ReferenceTracker returns potential references

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#12826]
**Tell us about your environment**

* **ESLint Version:** 7.0.0-alpha.0
* **Node Version:** v12.14.0
* **npm Version:** v6.13.4

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
    parserOptions: {
        ecmaVersion: 2020
    },
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

[Online Demo Link](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgcHJlZmVyLW9iamVjdC1zcHJlYWQ6IFwiZXJyb3JcIiAqL1xuXG52YXIgZG9Tb21ldGhpbmc7XG5cbmlmIChmb28pIHtcbiAgICBkb1NvbWV0aGluZyA9IE9iamVjdC5hc3NpZ247XG59IGVsc2Uge1xuICAgIGRvU29tZXRoaW5nID0gc29tZXRoaW5nRWxzZTtcbn1cblxudmFyIGJhciA9IGRvU29tZXRoaW5nKHt9LCBhLCBiKTtcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

```js
/*eslint prefer-object-spread: "error" */

var doSomething;

if (foo) {
    doSomething = Object.assign;
} else {
    doSomething = somethingElse;
}

var bar = doSomething({}, a, b);

```

```bash
eslint index.js --fix
```

**What did you expect to happen?**

I guess this shouldn't be reported and auto-fixed.

**What actually happened? Please include the actual, raw output from ESLint.**

Auto-fixed to:

```js
/*eslint prefer-object-spread: "error" */

var doSomething;

if (foo) {
    doSomething = Object.assign;
} else {
    doSomething = somethingElse;
}

var bar = { ...a, ...b};

```

**Are you willing to submit a pull request to fix this bug?**

Yes, though this would need some design first.

The problem is: `eslint-utils.ReferenceTracker` returns references that may be the targeted global in some code paths, but may be something else in other code paths. This feature might be not suitable for some rules.

**Comment by kaicataldo** [comment]
I have verified this bug. Agreed this should be fixed!

**Comment by kaicataldo** [comment]
Reopening this because this is a bug that results in the autofixer changing the execution of code.

**Comment by aggmoulik** [comment]
Hy @kaicataldo  I am looking into this issue. Can we explain a bit more how can I proceed ?

**Comment by mdjermanovic** [comment]
We could add an option in ReferenceTracker to return only certain references, or maybe change the rule to not use ReferenceTracker and manually find a smaller set of references.

@mysticatea what do you think about this?
